src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ActionAttributeRouteModel.cs

```diff
@@ -3,11 +3,141 @@
 
 using System.Collections.Generic;
 using System.Linq;
+using Microsoft.AspNetCore.Mvc.ActionConstraints;
+using Microsoft.AspNetCore.Mvc.Routing;
 
 namespace Microsoft.AspNetCore.Mvc.ApplicationModels
 {
     internal static class ActionAttributeRouteModel
     {
+        public static IEnumerable<SelectorModel> FlattenSelectors(ActionModel actionModel)
```

Comment:
This code got a little more complicated because it picked up responsibilities from CADP. There are no direct unit tests for this class, and I didn't start writing them in this PR. The tests for CADP cover these scenarios **IN DEPTH** and were very helpful in refactoring.